File: robot_ws/src/shaq_core/scripts_camera/hoop_detection_TFLite.cpp
#include "rclcpp/rclcpp.hpp"
#include "ament_index_cpp/get_package_share_directory.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include <cv_bridge/cv_bridge.hpp>
#include <opencv2/opencv.hpp>
#include <chrono>

#include "tensorflow/lite/interpreter.h"
#include "tensorflow/lite/kernels/register.h"
#include "tensorflow/lite/model.h"

using std::placeholders::_1;
using namespace std::chrono_literals;

class HoopDetectionNode : public rclcpp::Node
{
public:
    HoopDetectionNode()
    : Node("Hoop_Detection_TFLite"), x_(0.0), y_(0.0), led_state_(false)
    {
        std::string package_share_directory = ament_index_cpp::get_package_share_directory("shaq_core");
        std::string model_path = package_share_directory + "/models/trainvschair_int8.tflite";

        model_ = tflite::FlatBufferModel::BuildFromFile(model_path.c_str());
        if (!model_) {
            RCLCPP_FATAL(this->get_logger(), "Failed to load TFLite model");
            rclcpp::shutdown();
            return;
        }

        tflite::ops::builtin::BuiltinOpResolver resolver;
        tflite::InterpreterBuilder(*model_, resolver)(&interpreter_);
        if (!interpreter_) {
            RCLCPP_FATAL(this->get_logger(), "Failed to construct TFLite interpreter");
            rclcpp::shutdown();
            return;
        }

        if (interpreter_->AllocateTensors() != kTfLiteOk) {
            RCLCPP_FATAL(this->get_logger(), "Failed to allocate TFLite tensors");
            rclcpp::shutdown();
            return;
        }

        input_height_ = interpreter_->tensor(interpreter_->inputs()[0])->dims->data[1];
        input_width_ = interpreter_->tensor(interpreter_->inputs()[0])->dims->data[2];
        input_channels_ = interpreter_->tensor(interpreter_->inputs()[0])->dims->data[3];

        image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/shaq/image_raw", rclcpp::SensorDataQoS(),
            std::bind(&HoopDetectionNode::imageCallback, this, _1)
        );

        data_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/shaq/send_where_hoop", rclcpp::SensorDataQoS());
        led_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/shaq/led", 10);
        image_pub_ = this->create_publisher<sensor_msgs::msg::Image>("/shaq/image/hoop_annotated", 10);

        timer_ = this->create_wall_timer(50ms, std::bind(&HoopDetectionNode::publishData, this));
    }

private:
    void imageCallback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        try {
            cv::Mat frame = cv_bridge::toCvCopy(msg, "bgr8")->image;

            cv::Mat resized;
            cv::resize(frame, resized, cv::Size(input_width_, input_height_));
            resized.convertTo(resized, CV_32FC3, 1.0 / 255.0);
            cv::cvtColor(resized, resized, cv::COLOR_BGR2RGB);

            float* input = interpreter_->typed_input_tensor<float>(0);
            std::memcpy(input, resized.data, sizeof(float) * input_height_ * input_width_ * input_channels_);

            if (interpreter_->Invoke() != kTfLiteOk) {
                RCLCPP_ERROR(this->get_logger(), "TFLite inference failed");
                return;
            }

            float* output = interpreter_->typed_output_tensor<float>(0);
            float conf = output[4];

            if (conf > 0.2) {
                x_ = output[0];
                y_ = output[1];
                float w = output[2];
                float h = output[3];

                // Convert normalized (x, y, w, h) to pixel bounding box
                int x_pixel = static_cast<int>(x_);
                int y_pixel = static_cast<int>(y_);
                int width = static_cast<int>(w);
                int height = static_cast<int>(h);

                int x1 = x_pixel - width / 2;
                int y1 = y_pixel - height / 2;
                int x2 = x_pixel + width / 2;
                int y2 = y_pixel + height / 2;

                // Draw bounding box on the original image
                cv::rectangle(frame, cv::Point(x1, y1), cv::Point(x2, y2), cv::Scalar(0, 255, 0), 2);
                cv::putText(frame, "Hoop", cv::Point(x1, y1 - 10), cv::FONT_HERSHEY_SIMPLEX, 0.5, {255, 0, 0}, 1);
            } else {
                x_ = y_ = 0.0;
            }

            center_x_ = frame.cols / 2.0;
            center_y_ = frame.rows / 2.0;

            // Publish annotated image
            auto image_msg = cv_bridge::CvImage(msg->header, "bgr8", frame).toImageMsg();
            image_pub_->publish(*image_msg);

        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
        }
    }

    void publishData()
    {
        geometry_msgs::msg::Twist hoop_msg;
        geometry_msgs::msg::Twist led_msg;

        hoop_msg.linear.x = x_;
        hoop_msg.linear.y = y_;
        hoop_msg.angular.x = 270.0;
        hoop_msg.angular.y = center_y_;

        led_state_ = (x_ >= 265.0 && x_ <= 275.0);
        led_msg.linear.x = static_cast<double>(led_state_);

        data_pub_->publish(hoop_msg);
        led_pub_->publish(led_msg);
    }

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr data_pub_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr led_pub_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub_;
    rclcpp::TimerBase::SharedPtr timer_;

    double x_, y_, center_x_, center_y_;
    bool led_state_;

    std::unique_ptr<tflite::FlatBufferModel> model_;
    std::unique_ptr<tflite::Interpreter> interpreter_;

    int input_height_, input_width_, input_channels_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<HoopDetectionNode>());
    rclcpp::shutdown();
    return 0;
}

File: robot_ws/src/shaq_core/scripts_camera/hoop_detection_tflite.py
#!/usr/bin/env python3

import rclpy
from rclpy import qos
from rclpy.node import Node
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import os
import numpy as np
import cv2
import time
from ament_index_python.packages import get_package_share_directory
import tensorflow.lite as tflite  # or use `import tensorflow.lite as tflite` if using full TensorFlow

# Load TFLite model
package_share_directory = get_package_share_directory("shaq_core")
model_path = os.path.join(package_share_directory, "models", "trainvschair_float16.tflite")

interpreter = tflite.Interpreter(model_path=model_path)
interpreter.allocate_tensors()

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
input_size = (256, 256)


class mainRun(Node):
    def __init__(self):
        super().__init__("Hoop_Detection")

        self.bridge = CvBridge()

        self.x = 0.0
        self.y = 0.0
        self.center_x = 0.0
        self.center_y = 0.0
        self.led_state = False

        self.sent_where_hoop = self.create_publisher(
            Twist, "/shaq/send_where_hoop", qos_profile=qos.qos_profile_sensor_data
        )
        self.image_pub = self.create_publisher(
            Image, "/shaq/image/annotated_image", qos_profile=qos.qos_profile_sensor_data
        )
        self.sent_led = self.create_publisher(
            Twist, "/shaq/led", qos_profile=qos.qos_profile_default
        )
        self.subscription = self.create_subscription(
            Image,
            "/shaq/image_raw",
            self.image_callback,
            qos.qos_profile_sensor_data
        )

        self.sent_data_timer = self.create_timer(0.05, self.sendData)

    def image_callback(self, msg):
        start_time = time.time()
        try:
            frame = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
        except Exception as e:
            self.get_logger().error(f"CvBridge error: {e}")
            return

        frame_resized = cv2.resize(frame, input_size)
        h_resized, w_resized = frame_resized.shape[:2]

        # Prepare input for TFLite model
        input_tensor = frame_resized.astype(np.float32) / 255.0
        input_tensor = np.expand_dims(input_tensor, axis=0)  # NHWC
        if input_details[0]['dtype'] == np.uint8:
            input_tensor = (input_tensor * 255).astype(np.uint8)

        interpreter.set_tensor(input_details[0]['index'], input_tensor)
        interpreter.invoke()

        output = interpreter.get_tensor(output_details[0]['index'])  # shape: (N, 5)
        output = np.squeeze(output, axis=0).T  # shape (5, N)

        detections = [det for det in output if det[4] > 0.35]

        if detections:
            best_det = max(detections, key=lambda d: d[4])
            x, y, w, h, conf = best_det
            # print(conf)

            x1 = int((x - w / 2) * w_resized)
            y1 = int((y - h / 2) * h_resized)
            x2 = int((x + w / 2) * w_resized)
            y2 = int((y + h / 2) * h_resized)

            self.x = x * w_resized
            self.y = y * h_resized

            cv2.rectangle(frame_resized, (x1, y1), (x2, y2), (0, 255, 0), 2)
            cv2.putText(frame_resized, f"hoop: {conf:.2f}", (x1, max(y1 - 10, 0)),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 0), 2)
        else:
            self.x, self.y = 0.0, 0.0

        self.center_x = w_resized / 2.0
        self.center_y = h_resized / 2.0

        try:
            annotated_msg = self.bridge.cv2_to_imgmsg(frame_resized, encoding="bgr8")
            self.image_pub.publish(annotated_msg)
        except Exception as e:
            self.get_logger().error(f"Image publishing error: {e}")
        
        elapsed_ms = (time.time() - start_time) * 1000.0
        # print((time.time() - start_time) * 1000.0)
        self.get_logger().info(f"Inference time: {elapsed_ms:.2f} ms")

    def sendData(self):
        hoopdata_msg = Twist()
        led_msg = Twist()

        hoopdata_msg.linear.x = self.x
        hoopdata_msg.linear.y = self.y
        hoopdata_msg.angular.x = self.center_x
        hoopdata_msg.angular.y = self.center_y

        self.led_state = 265 <= self.x <= 275
        led_msg.linear.x = float(self.led_state)

        self.sent_led.publish(led_msg)
        self.sent_where_hoop.publish(hoopdata_msg)


def main():
    rclpy.init()
    node = mainRun()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
